Add options to control COCopier behavior and fix copying non-composite references

qmathe opened the pull request:
The 3 key changes are:
- support copying non-composite references to arbitrary depths (possibly copying the full source item graph)
- new copying behavior options
  - reuse existing source item UUIDs rather than creating new ones
    - can be used to easily clone some part of an item graph (e.g. a composite item tree) without changing its identity
  - choose between copy or alias for non-composite references that exist in the destination
  - choose between copy or ignore for non-composite references that are missing in the destination

ericwa commented on Core/COCopier.h:
```diff
@@ -11,6 +11,39 @@
+     */
+    COCopierCopiesNonCompositeReferencesMissingInDestination = 4,
+    /**
+     * Whether items reachable through non composite references are copied, when they exist in the
```
Copying items that exists in the destination means updating the item to match the state in the source item graph I guess?

qmathe replied:
It depends whether you pass COCopierReusesSourceUUID. 

If you pass COCopierReusesSourceUUID, then the item in the destination is updated to match the item in the source graph.

If you don't pass  COCopierReusesSourceUUID, then the corresponding item in the destination is left untouched, because the item in the source graph is copied into the destination with a different UUID.

I'll add a note about this point to make it clearer.

ericwa reviewed (approved):
I just skimmed the implementation but this sounds good!

qmathe commented:
I updated the API documentation based on your feedback and I also documented how the new options interacts with kCOIsSharedItemProperty.

In addition to this, I rewrote the copy algorithm to collect item UUIDs with a single traversal. The previous implementation was using two distinct recursive functions to collect composite references and non-composite references. The resulting code is shorter and should be faster.
